transformer_engine/pytorch/ops/basic/basic_linear.py

```diff
@@ -369,15 +380,21 @@ def _functional_forward(
         device = canonicalize_device(device)
         if device.type != "cuda":
             raise ValueError(f"Only CUDA devices are supported (got {device})")
+        if out is not None and not devices_match(out.device, device):
+            raise ValueError(
+                f"Output tensor has invalid device (expected {device}, got {out.device})"
+            )
 
         # Check datatype
         if dtype is None:
             dtype = weight.dtype
         dtype = canonicalize_dtype(dtype)
         if dtype not in (torch.float32, torch.float16, torch.bfloat16):
             raise ValueError(f"Supported dtypes are float32, float16, bfloat16 (got {dtype})")
+        if out is not None and out.dtype != dtype:
```

Comment:
When `dtype` is None and `out` is not, you should make sure that `dtype` becomes `out.dtype` rather than `weight.dtype`.